### ISSUE TITLE
proposal: indexBy

### ISSUE THREAD

**Issue by tjmehta** [tjmehta/101#73]
Should it work with reduce?
```js
[ 
  { foo: 'bar', bar: 'yo' }, 
  { foo: 'qux' }, 
  { foo: 'flux' }, 
  { foo: 100 }
].reduce(indexBy('foo'), {});
/* returns 
{
  'bar': { foo: 'bar', bar: 'yo' },
  'foo':  { foo: 'qux' }, 
  'flux': { foo: 'flux' },
  '100': { foo: 100 }
}
*/
```
Note: It requires an initial value which makes it a less "plug and play".

Or just 

```js
indexBy(arr, 'foo')
```

**Comment by tjmehta** [comment]
@stoeffel @tomekwi @bkendall @podviaznikov thoughts?

**Comment by tjmehta** [comment]
ooh maybe it should take a function:
```js
reduce(indexBy(pluck('foo')), {})
// or
indexBy(arr, pluck('foo'))
```

**Comment by stoeffel** [comment]
+1 for taking a function.
I think both `indexBy` and `groupBy` should take a function.

**Comment by tomekwi** [comment]
Great idea for a function!

> ```js reduce(indexBy(pluck('foo')), {}) ```

My favorite.

I guess it's ok for a reduce function to take an initial value. I use it almost every time I use reduce anyway – the results are much more predictable.

**Comment by tjmehta** [comment]
fixed in #99
